### PR TITLE
Curvature-Adaptive Surface Loss: Spatial High-Frequency Weighting

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The merged DCT frequency-weighted loss (PR #2184, p_tan -0.3%) forces the model to attend to high-frequency spectral components of the surface pressure profile. This spatial loss experiment pursues the same goal from the **spatial domain**: upweight the surface loss at high-curvature surface nodes (leading edge, trailing edge) where:

1. Pressure gradients are steepest — these are the hardest predictions
2. Errors compound most (LE stagnation point, TE separation)
3. A single mis-predicted node has high impact on integrated forces

**The curvature proxy** `curv = ||standardized_dsdf[0:4]|| * is_surface` (x channel 24) is already computed in the training loop and captures exactly this: it peaks at LE/TE (high gradient magnitude) and is near-zero at mid-chord. It's zero for volume nodes.

**The weight formula:**
```
weight(node) = 1.0 + alpha * (curv_normalized(node))
```
where `curv_normalized = curv / (max_curv_per_sample + eps)`, clamped to [0, 1]. This upweights LE/TE nodes by up to `(1 + alpha)x` while keeping mid-chord at `1.0x`. The total loss scale is approximately preserved.

**Why this is complementary to DCT freq loss (#2184):** DCT targets spectral high-frequency content on the full pressure profile. Curvature-adaptive loss targets spatially specific nodes. For cases where the model misses LE/TE pressure peaks but gets the mid-chord right (i.e., spatially localized errors), curvature weighting should help where DCT cannot.

**Kaggle competition analogy:** Sample-weighted loss based on difficulty is a standard winning technique. Here we apply it at the node level — the "hard samples" are the high-curvature surface nodes.

### Step 1: Add CLI flags

```python
# In TrainConfig dataclass:
curv_loss_weight: float = 0.0    # alpha: upweight curvature nodes (0 = disabled)
```

### Step 2: Implement per-node curvature weighting in surface loss

Find the surface loss computation in the training loop (~lines 1891–1914). After `abs_err` is computed, add curvature weights to the surface pressure MAE:

```python
if cfg.curv_loss_weight > 0.0:
    # curv is in extended x at channel index 24, shape [B, N, 1]
    # Already computed and appended: x = torch.cat([x_std, curv, dist_feat], dim=-1)
    curv_surf = x[:, :, 24:25]  # [B, N, 1], 0 for volume nodes
    
    # Normalize per sample: curv / max_curv (so max weight = 1 + alpha)
    max_curv = curv_surf.max(dim=1, keepdim=True).values.clamp(min=1e-6)  # [B, 1, 1]
    curv_norm = (curv_surf / max_curv).clamp(0.0, 1.0)  # [B, N, 1]
    
    # Per-node weight: 1.0 + alpha * curv_norm
    node_weights = 1.0 + cfg.curv_loss_weight * curv_norm  # [B, N, 1]
    
    # Apply to pressure MAE (channel 2) on surface nodes only
    # Replace abs_err[:, :, 2:3] with abs_err[:, :, 2:3] * node_weights in surf_loss computation
    surf_pres_weighted = abs_err[:, :, 2:3] * node_weights  # [B, N, 1]
else:
    surf_pres_weighted = abs_err[:, :, 2:3]  # unchanged

# Then use surf_pres_weighted instead of abs_err[:, :, 2:3] in the surf_loss lines
```

The weighting must be applied BEFORE the `surf_mask` accumulation and BEFORE the tandem boost. The tandem boost still applies on top of the per-node weighting.

Specifically, replace `abs_err[:, :, 2:3]` with `surf_pres_weighted` in the surface loss accumulation lines:

```python
# Original:
surf_per_sample = (abs_err[:, :, 2:3] * surf_mask.unsqueeze(-1)).sum(dim=(1,2)) \
                   / surf_mask.sum(dim=1).clamp(min=1).float()

# Modified:
surf_per_sample = (surf_pres_weighted * surf_mask.unsqueeze(-1)).sum(dim=(1,2)) \
                   / surf_mask.sum(dim=1).clamp(min=1).float()
```

Do NOT apply curvature weighting in the **validation loop** — validation should always be unweighted MAE for fair comparison across experiments.

### Step 3: Run 3 alpha values, 1 seed each + best gets 2nd seed

```bash
# alpha=0.5 (50% upweight at LE/TE), seed 42
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/curv-loss-a05-s42" \
  --wandb_group "askeladd/curv-adaptive-loss" \
  --curv_loss_weight 0.5 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# alpha=1.0 (100% upweight), seed 42 — same flags with --curv_loss_weight 1.0
# alpha=2.0 (200% upweight), seed 42 — same flags with --curv_loss_weight 2.0
```

Run all 3 configs with seed 42. The best config gets a second seed (73) for 2-seed validation.

**If curv_norm feels too noisy:** Instead of per-sample normalization, use a global normalization constant. Compute `global_max_curv` across all training samples on epoch 0 and cache it, then use `curv_norm = curv_surf / global_max_curv`. This is more stable but slightly more complex to implement.

### Key implementation notes

1. **curv = 0 for volume nodes** — the `is_surface` multiplication already zeroes out volume nodes in curv. So `node_weights = 1.0 + alpha * 0 = 1.0` for volume nodes. The volume loss is unaffected.

2. **Do NOT change validation loss** — only training loss gets curvature weighting. Validation always uses standard MAE for fair comparison.

3. **Verify with a sanity print:** In epoch 0, print the mean node_weight for surface nodes vs volume nodes. Surface mean should be > 1.0 (confirming upweighting), volume mean should be exactly 1.0.

4. **Monitor loss scale:** If curv_loss_weight is too large, the loss scale increases and may destabilize training. Watch for early divergence or loss exploding. If so, scale down alpha.

5. **VRAM:** Negligible overhead — no new parameters, no architecture change. Expect ~43 GB.

6. **If `curv` is noisy (not peaking at LE/TE as expected):** Log curv values for a few surface profiles. If the proxy doesn't reflect true curvature, try computing node-by-node arc-length curvature from adjacent surface nodes instead.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in | **13.21** | < 13.21 |
| p_oodc | **7.82** | < 7.82 |
| **p_tan** | **28.50** | **< 28.50** |
| p_re | **6.45** | < 6.45 |

**Baseline W&B runs:** 6yfv5lio (s42, p_tan=28.432), etepxvjc (s73, p_tan=28.572) — PR #2184 (DCT Freq Loss w=0.05)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2196 because Closing as duplicate — askeladd is already running the inter-foil distance feature experiment in PR #2195. The curvature-adaptive surface loss experiment is assigned to tanjiro in PR #2197, which has the final implementation. Avoiding redundant GPU usage..
